control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -1621,10 +1626,15 @@ func getMultiPortIdx(pod corev1.Pod, serviceEndpoints corev1.Endpoints) int {
 	return -1
 }
 
-func addressIsMissingFromEndpointsMap(address string, endpointsAddressesMap map[string]bool) bool {
-	if endpointsAddressesMap == nil {
+// deregister returns that the address is marked for deregistration if the map is nil or if the address is explicitly
+// marked in the map for deregistration.
+func deregister(address string, deregisterEndpointAddress map[string]bool) bool {
+	if deregisterEndpointAddress == nil {
 		return true
 	}
-	_, ok := endpointsAddressesMap[address]
-	return !ok
+	deregister, ok := deregisterEndpointAddress[address]
+	if ok {
+		return deregister
+	}
+	return false
```

Comment:
Don't we want to return `true` if it's not in the map, like in the case an service instance no longer appears in the endpoints subset?

Reply:
Thanks for this-- I intended the change to be that only if you explicitly add the address to the map for deregistration then you should deregister, but I think there's too many cases where we do rely on the implicit deregistration when a service instance no longer appears in the endpoints, and a bunch of tests were failing. So I just changed the logic to be very specific to the pod-fetch-error case, so only if there is a k8s api related pod fetch error, we explicitly choose to add it to the map and say "don't deregister this", and now the deregister function does return true if it's not in the map!